File: nqueens.py
import random


def is_valid(assignment):
    for col_i, row_i in enumerate(assignment):
        for col_j, row_j in enumerate(assignment[col_i + 1:], col_i + 1):
            if is_conflict(row_i, col_i, row_j, col_j):
                return False
    return True


def is_conflict(row1, col1, row2, col2):
    return (row1 == row2) or (abs(row1 - row2) == abs(col1 - col2))


def initialize_queens(N):
    r = 1
    queens = []
    for i in range(N):
        queens.append(r)
        r += 2
        if r >= N:
            r = 0
    return queens


def row_conflicts(queens: list, N, col):
    row_confs = [0 for _ in range(N)]
    for col_q, row_q in enumerate(queens):
        if col_q == col:
            continue
        row_confs[row_q] += 1  # add conflict for matching row

        if 0 <= row_q + abs(col_q - col) < N:
            row_confs[row_q + abs(col_q - col)] += 1
        if 0 <= row_q - abs(col_q - col) < N:
            row_confs[row_q - abs(col_q - col)] += 1
    return row_confs


def build_conflicts(queens: list):
    rows = {}
    rdiags = {}
    ldiags = {}
    N = len(queens)
    for col, row in enumerate(queens):
        rows[row] = rows.get(row, -1) + 1
        rdiags[row - col] = rdiags.get(row - col, -1) + 1
        ldiags[row - (N - col)] = ldiags.get(row - (N - col), -1) + 1
    conflicts = []
    for col, row in enumerate(queens):
        conflicts.append(rows.get(row, 0) + rdiags.get(row - col, 0) + ldiags.get(row - (N - col), 0))
    return conflicts


def update_conflicts(queens: list, conflicts: list, col, new_row):
    conflicts[col] = 0
    old_row = queens[col]

    for col2, row2 in enumerate(queens):
        if col == col2:
            continue
        # remove all conflicts with old_row
        if (old_row == row2) or (abs(old_row - row2) == abs(col - col2)):
            conflicts[col2] -= 1
        # add all conflicts with new_row
        if (new_row == row2) or (abs(new_row - row2) == abs(col - col2)):
            conflicts[col2] += 1
            conflicts[col] += 1


def pick_position(arr, N, cond):
    return random.choice([i for i in range(N) if cond(arr[i])])


def min_conflict(N, max_steps=10 ** 6):
    queens = initialize_queens(N)
    conflicts = build_conflicts(queens)  # O(n)
    for i in range(max_steps):  # O(N)

        if sum(conflicts) == 0:
            return queens, i

        col = pick_position(conflicts, N, lambda x: x > 0)  # O(N)
        row_confs = row_conflicts(queens, N, col)  # O(N)

        min_conf = min(row_confs)  # O(N)
        new_row = pick_position(row_confs, N, lambda x: x == min_conf)  # O(N)
        # update conflicts
        update_conflicts(queens, conflicts, col, new_row)  # O(N)
        queens[col] = new_row
    return queens, max_steps

File: nqueens.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <chrono>

typedef std::vector<int> vec;
typedef std::pair<vec, int> pair;

bool is_valid(vec &queens, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = i + 1; j < N; j++) {
            if (queens[i] == queens[j] || std::abs(queens[i] - queens[j]) == std::abs(i - j)) {
                return false;
            }
        }
    }
    return true;
}

vec initialize_queens(int N) {
    int r = 1;
    vec queens(N);

    for (int i = 0; i < N; i++) {
        queens[i] = r;
        r += 2;
        if (r >= N) {
            r = 0;
        }
    }
    return queens;
}

void row_conflicts(vec &queens, vec &row_conflicts, int N, int col) {
    for (int i = 0; i < N; i++) {
        row_conflicts[i] = 0;
    }
    for (int i = 0; i < N; i++) {
        if (i == col) {
            continue;
        }
        row_conflicts[queens[i]] += 1;

        int rd = queens[i] + std::abs(i - col);
        int ld = queens[i] - std::abs(i - col);

        if (rd >= 0 && rd < N) {
            row_conflicts[rd] += 1;
        }
        if (ld >= 0 && ld < N) {
            row_conflicts[ld] += 1;
        }
    }
}

vec build_conflicts(vec &queens, int N) {
    vec rows(N, 0);
    vec rdiags(2 * N, 0);
    vec ldiags(2 * N, 0);

    for (int i = 0; i < N; i++) {
        int row = queens[i];
        rows[row] += 1;
        rdiags[(N - 1) + row - i] += 1;
        ldiags[(N - 1) + row - (N - i - 1)] += 1;
    }

    vec conflicts(N, 0);
    for (int i = 0; i < N; i++) {
        int row = queens[i];
        conflicts[i] = (rows[row] - 1) + (rdiags[(N - 1) + row - i] - 1) + (ldiags[(N - 1) + row - (N - i - 1)] - 1);
    }

    return conflicts;
}

int random_position_greater_zero(vec &list, int N, double rand) {
    vec possible;

    for (int i = 0; i < N; i++) {
        if (list[i] > 0) {
            possible.push_back(i);
        }
    }

    int rand_i = (int) (rand * possible.size());
    return possible[rand_i];
}

int random_position_equal_to(vec &list, int value, int N, double rand) {
    vec possible;

    for (int i = 0; i < N; i++) {
        if (list[i] == value) {
            possible.push_back(i);
        }
    }

    int rand_i = (int) (rand * possible.size());
    return possible[rand_i];
}


int update_conflicts(vec &queens, vec &conflicts, int N, int col, int new_row) {
    int old_row = queens[col];
    conflicts[col] = 0;

    int total_conflicts = 0;

    for (int i = 0; i < N; i++) {
        if (i == col) {
            continue;
        }
        int row = queens[i];
        if ((old_row == row) || std::abs(old_row - row) == std::abs(col - i)) {
            conflicts[i] -= 1;
        }

        if ((new_row == row) || std::abs(new_row - row) == std::abs(col - i)) {
            conflicts[i] += 1;
            conflicts[col] += 1;
        }
        total_conflicts += conflicts[i];
    }

    return total_conflicts + conflicts[col];;
}

int sum(vec &list, int N) {
    int sum = 0;
    for (int i = 0; i < N; i++) {
        sum += list[i];
    }
    return sum;
}

int min(vec &list, int N) {
    int min = INT_MAX;
    for (int i = 0; i < N; i++) {
        if (list[i] < min) {
            min = list[i];
        }
    }
    return min;
}

pair iterative_repair(int N) {
    const int MAX_STEPS = 1000000;

    vec queens = initialize_queens(N);
    vec conflicts = build_conflicts(queens, N);
    vec row_confs(N, 0);

    std::random_device rd;
    std::mt19937 e2(rd());
    std::uniform_real_distribution<> dist(0, 1);
    int total_conflicts = sum(conflicts, N);

    for (int i = 0; i < MAX_STEPS; i++) {
        if (total_conflicts == 0) {
            return pair(queens, i);
        }

        int col = random_position_greater_zero(conflicts, N, dist(e2));
        row_conflicts(queens, row_confs, N, col);
        int min_conflict = min(row_confs, N);
        int new_row = random_position_equal_to(row_confs, min_conflict, N, dist(e2));
        
        total_conflicts = update_conflicts(queens, conflicts, N, col, new_row);
        queens[col] = new_row;
    }

    return pair(queens, MAX_STEPS);
}

int user_input_N() {
    int N;
    std::cout << "N: ";
    std::cin >> N;
    return N;
}

int main() {
    int N = 0;

    while (N != -1) {
        N = user_input_N();
        auto start = std::chrono::high_resolution_clock::now();
        auto result = iterative_repair(N);
        auto stop = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::seconds>(stop - start);


        std::cout << N << "\t" << result.second << "\t" << "took " << duration.count() << "s" << std::endl;

        start = std::chrono::high_resolution_clock::now();
        bool valid = is_valid(result.first, N);
        stop = std::chrono::high_resolution_clock::now();
        duration = std::chrono::duration_cast<std::chrono::seconds>(stop - start);

        char *s = const_cast<char *>(valid ? "valid" : "invalid");
        std::cout << "and " << s << "! in " << duration.count() << "s" << std::endl;
    }
    return 0;
}
